CesiumAsync/include/CesiumAsync/SharedFuture.h

```diff
@@ -0,0 +1,254 @@
+   * return value of the catch callback.
+   *
+   * @tparam Func The type of the function.
+   * @param f The function.
+   * @return A future that resolves after the supplied function completes.
+   */
+  template <typename Func> Future<T> catchInMainThread(Func&& f) {
+    return this->catchWithScheduler(
+        this->_pSchedulers->mainThread.immediate,
+        std::forward<Func>(f));
+  }
+
+  /**
+   * @brief Registers a continuation function to be invoked immediately, and
+   * invalidates this Future.
+   *
+   * When this Future is rejected, the continuation function will be invoked
+   * in whatever thread does the rejection. Similarly, if the Future is already
+   * rejected when `catchImmediately` is called, the continuation function will
+   * be invoked immediately before this method returns.
+   *
+   * If the function itself returns a `Future`, the function will not be
+   * considered complete until that returned `Future` also resolves.
+   *
+   * Any `then` continuations chained after this one will be invoked with the
+   * return value of the catch callback.
+   *
+   * @tparam Func The type of the function.
+   * @param f The function.
+   * @return A future that resolves after the supplied function completes.
+   */
+  template <typename Func> Future<T> catchImmediately(Func&& f) {
+    return this->catchWithScheduler(
+        async::inline_scheduler(),
+        std::forward<Func>(f));
+  }
+
+  /**
+   * @brief Waits for the future to resolve or reject and returns the result.
+   *
+   * This method must not be called from the main thread, the one that calls
+   * {@link AsyncSystem::dispatchMainThreadTasks}. Doing so can lead to a
+   * deadlock because the main thread tasks will never complete while this
+   * method is blocking the main thread.
+   *
+   * @return The value if the future resolves successfully.
+   * @throws An exception if the future rejected.
+   */
+  T wait() { return this->_task.get(); }
+
+private:
+  SharedFuture(
+      const std::shared_ptr<Impl::AsyncSystemSchedulers>& pSchedulers,
+      async::shared_task<T>&& task) noexcept
+      : _pSchedulers(pSchedulers), _task(std::move(task)) {}
+
+  template <typename Func, typename Scheduler>
+  Impl::ContinuationFutureType_t<Func, T>
+  thenWithScheduler(Scheduler& scheduler, const char* tracingName, Func&& f) {
+    // It would be nice if tracingName were a template parameter instead of a
+    // function parameter, but that triggers a bug in VS2017. It was previously
+    // a bug in VS2019, too, but has been fixed there:
+    // https://developercommunity.visualstudio.com/t/internal-compiler-error-when-compiling-a-template-1/534210
+#if CESIUM_TRACING_ENABLED
+    // When tracing is enabled, we measure the time between scheduling and
+    // dispatching of the work.
+    auto task = this->_task.then(
+        async::inline_scheduler(),
+        Impl::WithTracing<T>::begin(tracingName, std::forward<Func>(f)));
```

Comment:
this code doesn't seem to compile when enabling tracing. I think it's because `Impl::WithTracing<T>::begin()` returns a function with rvalue reference, but `shared_task` seems to only accepts value or const lvalue&